### PR TITLE
Simplify and fix show endpoint summary processing

### PR DESCRIPTION
**Pull request by robbrockbank**

### CODE REVIEW

File: calico_containers/calico_ctl/endpoint.py [hunk 1]
```diff
@@ -138,44 +138,21 @@ def endpoint_show(hostname, orchestrator_id, workload_id, endpoint_id,
...
-
-            # Add up how many endpoints each workload on this host has
-            num_endpoints = 0
-            for workload, endpoints in iter(table_dict[host].items()):
-                num_endpoints += endpoints
-
+        # For each host/orchestrator, keep track of unique workload IDs and
+        # total number of endpoints.  For simplicity we just store the working
+        # data in a list [set([workload IDs...], int(running total # eps)]
+        host_orch = {}
```
Comment (tomdee) [inline calico_containers/calico_ctl/endpoint.py]: I think expanding some of these short variable names would make this code more readable

File: calico_containers/calico_ctl/endpoint.py [hunk 2]
```diff
@@ -138,44 +138,21 @@ def endpoint_show(hostname, orchestrator_id, workload_id, endpoint_id,
...
+        for (hn, oi), data in host_orch.iteritems():
             # Add the results to this table
-            new_row = [endpoint.hostname,
-                       endpoint.orchestrator_id,
-                       num_workloads,
-                       num_endpoints]
-
-            x.add_row(new_row)
+            x.add_row([hn, oi, len(data[0]), data[1]])
+
```
Comment (tomdee) [inline calico_containers/calico_ctl/endpoint.py]: Does this mean you can get rid of vividict?

File: calico_containers/calico_ctl/endpoint.py [hunk 3]
```diff
@@ -138,44 +138,21 @@ def endpoint_show(hostname, orchestrator_id, workload_id, endpoint_id,
...
-            num_workloads = len(table_dict[host])
-
-            # Add up how many endpoints each workload on this host has
-            num_endpoints = 0
-            for workload, endpoints in iter(table_dict[host].items()):
-                num_endpoints += endpoints
-
+        # For each host/orchestrator, keep track of unique workload IDs and
+        # total number of endpoints.  For simplicity we just store the working
+        # data in a list [set([workload IDs...], int(running total # eps)]
```
Comment (tomdee) [inline calico_containers/calico_ctl/endpoint.py]: This isn't really a list though, it's a tuple - won't it always have two values? So a named tuple would probably result in clearer code here...

File: calico_containers/calico_ctl/endpoint.py (hunk 2 above)
Comment (robbrockbank) [inline calico_containers/calico_ctl/endpoint.py]: good spot!  - yes it means we can.

File: calico_containers/calico_ctl/endpoint.py [hunk 4]
```diff
@@ -138,44 +138,21 @@ def endpoint_show(hostname, orchestrator_id, workload_id, endpoint_id,
...
+        host_orch = {}
+        for ep in endpoints:
+            key = (ep.hostname, ep.orchestrator_id)
+            data = host_orch.setdefault(key, [set(), 0])
```
Comment (tomdee) [inline calico_containers/calico_ctl/endpoint.py]: data is too generic - can you think of a better name?

**Comment by tomdee** [comment]
Looks good but I think the code could be clearer. Can I take another look after you address the comments.

File: calico_containers/calico_ctl/endpoint.py (hunk 3 above)
Comment (robbrockbank) [inline calico_containers/calico_ctl/endpoint.py]: i could use a tuple, but they are immutable, so I'd have to create a new tuple to increment the count... or wrap it in a holder class.  Didn't feel like it was worth it being only a few lines of code.

**Comment by robbrockbank** [comment]
As discussed with Tom, beefed up general comment about summary processing.  Merging.

**Pull request merged**